File: algotithms/cpp/42. 接雨水.cpp
class Solution {
public:
    int trap(vector<int>& h) {
        int n = h.size();
        vector<int> left_max(n), right_max(n);
        left_max[0] = h[0], right_max[n-1] = h[n-1];
        for (int i = 1; i < n; i++) {
            left_max[i] = max(left_max[i-1], h[i]);
        }
        for (int i = n-2; i >= 0; i--) {
            right_max[i] = max(right_max[i+1], h[i]);
        }
        int res = 0;
        for (int i = 0; i < n; i++) {
            res += min(left_max[i], right_max[i]) - h[i];
        }
        return res;
    }
};

File: algotithms/python/42. 接雨水.py
class Solution:
    def trap(self, height: List[int]) -> int:
        n = len(height)
        pre_max = [0] * n
        pre_max[0] = height[0]
        for i in range(1, n):
            pre_max[i] = max(pre_max[i-1], height[i])
        
        suf_max = [0] * n
        suf_max[-1] = height[-1]
        for i in range(n-2, -1, -1):
            suf_max[i] = max(suf_max[i+1], height[i])
        
        ans = 0
        for h, pre, suf in zip(height, pre_max, suf_max):
            ans += min(pre, suf) - h
        return ans
        